### PR TITLE
Align image collection of extensions with existing pattern

### PR DESCRIPTION
**Pull request by rohitsakala**
## Issue: 

fixes https://github.com/rancher/rancher/issues/43779
 
## Problem

The [GetImages](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/resolve.go#L64) func is indirectly being called by a codebase in the catalogv1 package in airgap scenario.

[GetImages](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/resolve.go#L64) =>calls=> [AddImagesToImageListConfigMap](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/resolve.go#L109) =>calls=>[CatalogV1 Package](https://github.com/rancher/rancher/blob/release/v2.9/pkg/catalog/manager/catalog_sync.go#L123)

The [GetImages](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/resolve.go#L64) func is also called by [utility package](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/utilities/utilities.go#L144) for generating rancher-images.txt file for airgap scenarios.

**GetImages Func**
It takes in an argument `exportConfig` which defines where to find packages from? 

When called from [CatalogV2](https://github.com/rancher/rancher/blob/release/v2.9/pkg/catalog/manager/catalog_sync.go#L123) , it seems like it defines internal directories to fetch from. 

When called from [Utilities](https://github.com/rancher/rancher/blob/release/v2.9/pkg/image/utilities/utilities.go#L144), it defines Github URL's such as github.com/rancher/charts or github.com/rancher/system-charts. 

The [PR](https://github.com/rancher/rancher/pull/42565) adds a new entry for [Extensions](https://github.com/rancher/rancher/pull/42565/files#diff-df058c3539a16a7aa843e7dd88ea2ffb4c50e5ec45f91ab515b1a85d521748e6R82) which doesn't consider exportConfig and always fetches from github.com/rancher/ui-plugin-charts. 

 
## Solution

This PR now tries to use export config for [extensions](https://github.com/rancher/rancher/pull/43825/files#diff-2623863f3d8249bed94fa69df203ba439413583a04367f68031f176af5187200R138) as well when called from [utilities](https://github.com/rancher/rancher/pull/43825/files#diff-2623863f3d8249bed94fa69df203ba439413583a04367f68031f176af5187200R143)

### Manual Testing

Tested if it fixes the bug. 

### Automated Testing

- Unit - Added Unit tests for the GetImage function with respective to Extensions being set and not set.

## QA Testing Considerations

**Non Airgap Scenario**
- All Rancher test cases 

**AirGap Scenario**
- All Rancher test cases

**Rancher.images.txt Generation**
- Check if rancher-images.txt is generated correctly and has right entries.

### CODE REVIEW

**Review by jordojordo — APPROVED** [review]
LGTM - Thanks for fixing this @rohitsakala

File: pkg/image/resolve_test.go
Comment (snasovich) [inline pkg/image/resolve_test.go:221]: It may not be a good idea to add this unit test as per my understanding when it runs it actually makes HTTP calls to `https://api.github.com/repos/rancher/ui-plugin-charts/releases` which by definition makes it not a unit test as there is a dependency on external system outside of immediate `rancher/rancher` codebase. We may also get rate-limited by GH if this test runs a lot.

Is this something we already do in other unit tests in `rancher/rancher`?

Mocking HTTP server may be a way to go but not sure how involved it will be.

(The first test case is fine though)

**Review by snasovich — CHANGES_REQUESTED** [review]
@rohitsakala , thank you for updating the description - it helps.

I now see that in the case that was previously failing (https://github.com/rancher/rancher/blob/e0543ecbda992641787b3ecf0a0febbbabfce2d2/pkg/image/resolve.go#L109) the configuration will have empty `GithubEndpoints` so iteration in `FetchExtensionImages` (https://github.com/rancher/rancher/blob/e0543ecbda992641787b3ecf0a0febbbabfce2d2/pkg/image/extensions.go#L29) won't reach out to any HTTP endpoints.

I do have one concern with unit tests so keeping "Request changes" for now.

**Comment by snasovich** [comment]
@rohitsakala , additionally, `check-changes` PR [check](https://github.com/rancher/rancher/actions/runs/7395723999/job/20147443300) is failing. Please look into it too.

File: pkg/image/resolve_test.go
Comment (rohitsakala) [reply pkg/image/resolve_test.go]: Done. I have resolved it by using setupTestServer which was already used in other unit tests.

**Comment by rohitsakala** [comment]
> @rohitsakala , additionally, `check-changes` PR [check](https://github.com/rancher/rancher/actions/runs/7395723999/job/20147443300) is failing. Please look into it too.

This PR https://github.com/rancher/rancher/pull/43903 might fix it.

**Comment by snasovich** [comment]
> > @rohitsakala , additionally, `check-changes` PR [check](https://github.com/rancher/rancher/actions/runs/7395723999/job/20147443300) is failing. Please look into it too.
> 
> This PR #43903 might fix it.

Ah, the "New Year's `go generate`" issue. Happy New Year!

File: pkg/image/resolve_test.go
Comment (snasovich) [inline pkg/image/resolve_test.go:256]: Minor nit, but I think we can update iteration to use values rather than indexes for a bit cleaner code. We could also break as soon as the match is found. Same can be done for the iteration below.

Maybe there is something I'm missing why we have to use indexes here.

```suggestion
			for _, expected := range tt.expected {
				found := false
				for _, image := range imagesList {
					if strings.Contains(image, expected) {
						found = true
						break
					}
				}
				assertlib.True(t, found)
			}
```

**Review by snasovich — DISMISSED** [review]
LGTM, just one minor nit that is optional to address (don't want it to block merging if there is another approval - feel free to still count my approval if addressing this comment is the only change pushed after this approval).

File: pkg/image/resolve_test.go
Comment (snasovich) [reply pkg/image/resolve_test.go]: Perfect, thank you!

**Review by snasovich — APPROVED** [review]

**Review by HarrisonWAffel — APPROVED** [review]

**Review by diogoasouza — APPROVED** [review]
LGTM

**Pull request merged**
